include/circt/Dialect/Moore/MoorePasses.td

```diff
@@ -29,4 +29,40 @@ def SimplifyProcedures : Pass<"simplify-procedures", "moore::SVModuleOp"> {
     let constructor = "circt::moore::createSimplifyProceduresPass()";
 }
 
+def Dedup : Pass<"moore-dedup", "mlir::ModuleOp"> {
+  let summary = "Deduplicate modules which are generated by same instance in SV";
+  let description = [{
+    This pass detects modules which are structurally equivalent and removes the
+    duplicate module by replacing all instances of one with the other.
+    This pass is only suitable for software simulation. Different symbolName
+    modules relate to different hardware units in hardware simulation. So this pass
+    is not recommended in hardware simulation.
+    For example,
+    ```
+    module {
+    moore.module @top() {
+      %a = moore.net wire : <l4>
+      %0 = moore.read %a : l4
+      moore.instance "insA" @NestedA(a: %0: !moore.l4) -> ()
+      %1 = moore.read %a : l4
+      --- moore.instance "insB" @NestedA_0(a: %1: !moore.l4) -> ()
+      +++ moore.instance "insB" @NestedA(a: %1: !moore.l4) -> ()
+      moore.output
+      }
+      --- moore.module @NestedA_0(in %a : !moore.l4) {
+      ---   %a_0 = moore.net name "a" wire : <l4>
+      ---   moore.assign %a_0, %a : l4
+      ---   moore.output
+      --- }
+      moore.module @NestedA(in %a : !moore.l4) {
+      %a_0 = moore.net name "a" wire : <l4>
+      moore.assign %a_0, %a : l4
+      moore.output
+      }
+    }
+    ```
```

Comment:
Is it necessary to list examples of this?

Reply:
You are right. I may remove this.